src/basicfuns.jl

```diff
@@ -153,8 +153,37 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    t = log1p(exp(-abs(x)))
+    return x ≤ 0 ? t : t + x
+end
```

Comment:
What's the reason for this change? I am worried this will reduce accuracy eg for `ForwardDiff.Dual`.

Reply:
The bounds given before are specific to Float64, and should be inaccurate for other types. I prefer a more robust method as fallback. Note that this implementation is very accurate as far as I can tell, it's just slower.

Reply:
The point is that this fallback will be hit now whereas before a, for 64bit floating point numbers, more optimized version was used. And this will affect eg. `ForwardDiff.Dual`. The problem is not the accuracy of the Float64/32/16 implementations but the reduced accuracy for other types.

Reply:
My suggestion is to not touch the default definition in this PR since it requires caeeful evaluation for which types accuracy and performance would be reduced, and for which maybe it would be improved.

Reply:
I'm not sure I follow. `log1pexp(x::Real)` is the most accurate version, it is just slow.

So there can be an issue that `log1pexp(::ForwardDiff.Dual)` will now be slower. Is this your concern?

If that's so I agree. But then the point is that the fast implementation is using bounds tailored for Float64. So we can't just have that method be the fallback for all `Real`s, right?

Reply:
We want to move to this fallback definition, I think - but there are cases where the current fallback is accurate and fast, even though inputs are not Float64. And I don't think it's acceptable to introduce performance regressions in these cases, and in particular not in a PR where it is not necessary to change the fallback definition.

I suggest keeping it for now, opening an issue, evaluating which types are affected by the change, and then e.g. overloading `log1pexp` in ForwardDiff if needed, and finally replacing the fallback.

Reply:
I would suggest just special-casing `ForwardDiff.Dual` to avoid a regression (ie define a method, use bounds for underlying type).

Reply:
I'm marking this as resolved since it refers to an older version of the code.